client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/orderedgrouped/RssMergeManager.java

```diff
@@ -0,0 +1,606 @@
+    // (usedMemory + requestedSize > memoryLimit). When this thread is done
+    // fetching, this will automatically trigger a merge thereby unlocking
+    // all the stalled threads
+
+    if (usedMemory > memoryLimit) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(
+            srcAttemptIdentifier
+                + ": Stalling shuffle since usedMemory ("
+                + usedMemory
+                + ") is greater than memoryLimit ("
+                + memoryLimit
+                + ")."
+                + " CommitMemory is ("
+                + commitMemory
+                + ")");
+      }
+      return stallShuffle;
+    }
+
+    // Allow the in-memory shuffle to progress
+    if (LOG.isDebugEnabled()) {
+      LOG.debug(
+          srcAttemptIdentifier
+              + ": Proceeding with shuffle since usedMemory ("
+              + usedMemory
+              + ") is lesser than memoryLimit ("
+              + memoryLimit
+              + ")."
+              + "CommitMemory is ("
+              + commitMemory
+              + ")");
+    }
+    return unconditionalReserve(srcAttemptIdentifier, requestedSize, true);
+  }
+
+  private synchronized MapOutput unconditionalReserve(
+      InputAttemptIdentifier srcAttemptIdentifier, long requestedSize, boolean primaryMapOutput)
+      throws IOException {
+    usedMemory += requestedSize;
+    return MapOutput.createMemoryMapOutput(
+        srcAttemptIdentifier, this, (int) requestedSize, primaryMapOutput);
+  }
+
+  @Override
+  public synchronized void unreserve(long size) {
+    usedMemory -= size;
+    if (LOG.isDebugEnabled()) {
+      LOG.debug(
+          "Notifying unreserve : size="
+              + size
+              + ", commitMemory="
+              + commitMemory
+              + ", usedMemory="
+              + usedMemory
+              + ", mergeThreshold="
+              + mergeThreshold);
+    }
+    notifyAll();
+  }
+
+  @Override
+  public synchronized void releaseCommittedMemory(long size) {
+    commitMemory -= size;
+    unreserve(size);
+  }
+
+  @Override
+  public synchronized void closeInMemoryFile(MapOutput mapOutput) {
+    inMemoryMapOutputs.add(mapOutput);
+    trackAndLogCloseInMemoryFile(mapOutput);
+    commitMemory += mapOutput.getSize();
+    if (commitMemory >= mergeThreshold) {
+      startMemToRemoteMerge();
+    }
+  }
+
+  private void trackAndLogCloseInMemoryFile(MapOutput mapOutput) {
+    statsInMemTotal.updateStats(mapOutput.getSize());
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug(
+          "closeInMemoryFile -> map-output of size: "
+              + mapOutput.getSize()
+              + ", inMemoryMapOutputs.size() -> "
+              + inMemoryMapOutputs.size()
+              + ", commitMemory -> "
+              + commitMemory
+              + ", usedMemory ->"
+              + usedMemory
+              + ", mapOutput="
+              + mapOutput);
+    } else {
```

Comment:
```    
if (LOG.isDebugEnabled()) {
      LOG.debug(
          "closeInMemoryFile -> map-output of size: "
              + mapOutput.getSize()
              + ", inMemoryMapOutputs.size() -> "
              + inMemoryMapOutputs.size()
              + ", commitMemory -> "
              + commitMemory
              + ", usedMemory ->"
              + usedMemory
              + ", mapOutput="
              + mapOutput);
    }
statsInMemLastLog.updateStats(mapOutput.getSize());
```
Maybe we could remove the `else`.

Reply:
> ```
> if (LOG.isDebugEnabled()) {
>       LOG.debug(
>           "closeInMemoryFile -> map-output of size: "
>               + mapOutput.getSize()
>               + ", inMemoryMapOutputs.size() -> "
>               + inMemoryMapOutputs.size()
>               + ", commitMemory -> "
>               + commitMemory
>               + ", usedMemory ->"
>               + usedMemory
>               + ", mapOutput="
>               + mapOutput);
>     }
> statsInMemLastLog.updateStats(mapOutput.getSize());
> ```
> 
> Maybe we could remove the `else`.

@jerqi 
Here is to keep the same logic as tez.
In order to avoid excessive logs, info-level logs are printed every 30 seconds.

Reply:
OK.